### PR TITLE
Support different datatype between particle attributes and fluid data

### PR DESCRIPTION
**Pull request by koarakawaii**
* By specifying `FLOAT8_PAR` in `Makefile`, this PR supports using different datatype precision for particle attributes from that of fluid data
* Convert `real` to `real_par` related to particle attributes in all source files
* Use number of bytes instead of number of elements for both `LB_GetBufferData_MemAllocate_Send/LB_GetBufferData_MemAllocate_Recv`; both functions now return `void` pointer, and user needs to convert the `void` pointer to pointer with desired datatype manually

### CODE REVIEW

**Comment by koarakawaii** [comment]
@hyschive : I think this PR is ready to be reviewed~

**Comment by hsinhaoHHuang** [comment]
@hyschive  I've finished the tests for this PR.
-  I ran the `Plummer` test problem
    - with `OPT__INIT=3` and `PAR_INIT=3` (i.e., using `UM_IC` and `PAR_IC` for initialization).
-  All 16 combinations of  [`FLOAT8`, `FLOAT8_PAR`, `OPT__UM_IC_FLOAT8`, `PAR_IC_FLOAT8`] are tested.
- The images of projection after evolution all look the same (and also the same as the results I ran in the `main` branch with `OPT__INIT=1` and `PAR_INIT=1`).

**Comment by hyschive** [comment]
@hsinhaoHHuang To be more cautious, could you also test different combinations of `FLOAT8, FLOAT8_PAR` during restart? For example, you can have `FLOAT8=0, FLOAT8_PAR=1` at the beginning but `FLOAT8=1, FLOAT8_PAR=0` during restart.

**Comment by hsinhaoHHuang** [comment]
@hyschive I restarted my previous runs of Plummer from `Data_000010` and ran to the end.

- All 16 different combinations of 
[ `FLOAT8` at the beginning, `FLOAT8_PAR` at the beginning, `FLOAT8` during restart, `FLOAT8_PAR` during restart ] are tested.
- The images of projection after evolution also all look the same as before.

**Comment by hyschive** [comment]
@hsinhaoHHuang FTR. Please _approve_ this PR. Thanks.

**Review by hsinhaoHHuang — APPROVED** [review]

File: src/Particle/Par_UpdateTracerParticle.cpp [hunk 1]
```diff
@@ -143,16 +143,16 @@ void Par_UpdateTracerParticle( const int lv, const double TimeNew, const double
...
 //             determine time-step
-               dt = (real)TimeNew - ParTime[ParID];
+               dt = (real)TimeNew - (real)ParTime[ParID];
 
 //             convert time-step for comoving
 #              ifdef COMOVING
-               if ( ParTime[ParID] == (real)TimeOld )    dt = dt_com;   // avoid redundant calculations
-               else                                      dt = Mis_dTime2dt( ParTime[ParID], dt );
+               if ( ParTime[ParID] == (real_par)TimeOld )    dt = dt_com;   // avoid redundant calculations
+               else                                          dt = Mis_dTime2dt( (const double)ParTime[ParID], (const double)dt );
```
Comment (hyschive) [inline src/Particle/Par_UpdateTracerParticle.cpp:151]: Remove `const`

File: src/Particle/Par_UpdateTracerParticle.cpp [hunk 2]
```diff
@@ -143,16 +143,16 @@ void Par_UpdateTracerParticle( const int lv, const double TimeNew, const double
...
 #              endif
 
 //             predict the positions at TimeNew
-               for (int d=0; d<3; d++)    InterpParPos[d][p] = ParPos[d][ParID] + dt*ParVel[d][ParID];
+               for (int d=0; d<3; d++)    InterpParPos[d][p] = ParPos[d][ParID] + (real_par)dt*ParVel[d][ParID];
```
Comment (hyschive) [inline src/Particle/Par_UpdateTracerParticle.cpp:155]: Remove `(real_par)` if you agree upon declaring `dt` as `real_par`.

File: src/Particle/Par_UpdateTracerParticle.cpp [hunk 3]
```diff
@@ -193,24 +193,24 @@ void Par_UpdateTracerParticle( const int lv, const double TimeNew, const double
...
 //             determine time-step
-               dt = (real)TimeNew - ParTime[ParID];
+               dt = (real)TimeNew - (real)ParTime[ParID];
 
 //             convert time-step for comoving
 #              ifdef COMOVING
-               if ( ParTime[ParID] == (real)TimeOld )    dt = dt_com;   // avoid redundant calculations
-               else                                      dt = Mis_dTime2dt( ParTime[ParID], dt );
+               if ( ParTime[ParID] == (real_par)TimeOld )    dt = dt_com;   // avoid redundant calculations
+               else                                          dt = Mis_dTime2dt( (const double)ParTime[ParID], (const double)dt );
```
Comment (hyschive) [inline src/Particle/Par_UpdateTracerParticle.cpp:208]: Remove `const`

File: src/Particle/Par_UpdateTracerParticle.cpp [hunk 4]
```diff
@@ -193,24 +193,24 @@ void Par_UpdateTracerParticle( const int lv, const double TimeNew, const double
...
 #              endif
 
                for (int d=0; d<3; d++)
                   InterpParPos[d][p] = ParPos[d][ParID] +
-                     (real)0.5*dt*( Vel_Temp[d][p] + ParVel[d][ParID] );
+                     (real_par)0.5*(real_par)dt*( Vel_Temp[d][p] + ParVel[d][ParID] );
```
Comment (hyschive) [inline src/Particle/Par_UpdateTracerParticle.cpp:213]: Remove `(real_par)` in front of `dt` if you agree upon declaring `dt` as `real_par`.

File: src/Output/Output_DumpData_Total.cpp [hunk 5]
```diff
@@ -53,6 +53,14 @@ void Output_DumpData_Total( const char *FileName )
 
 
    if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s (DumpID = %d)     ...\n", __FUNCTION__, DumpID );
+#  ifdef PARTICLE
+   if ( (OPT__OUTPUT_TOTAL == OUTPUT_FORMAT_CBINARY) && (MPI_Rank == 0) )
```
[the remark this answers is not in the document]
Comment (koarakawaii) [reply src/Output/Output_DumpData_Total.cpp:57]: @hyschive Isn't this sanity check for identical floating-point precision between field and particle data (in general they can be different, if we dump snapshot as `HDF5`) for binary format snapshot output? So I am not sure why we can ignore the `OPT__OUTPUT_TOTAL == OUTPUT_FORMAT_CBINARY` condition. And the `MPI_Rank == 0` is for only calling `Aux_Error` once I think.

[the remark this answers is not in the document]
Comment (koarakawaii) [reply src/Output/Output_DumpData_Total.cpp:57]: @hyschive Oh, I figure it out. This `cpp` file itself is already for binary format output already, right?

[the remark this answers is not in the document]
Comment (hyschive) [reply src/Output/Output_DumpData_Total.cpp:57]: @koarakawaii Yes

[the remark this answers is not in the document]
Comment (hyschive) [reply src/Output/Output_DumpData_Total.cpp:57]: To be more cautious, we let all MPI ranks call `Aux_Error()` (but not `Aux_Message()`)

File: src/YT/YT_GetParticleAttribute.cpp
[the remark this answers is not in the document]
Comment (koarakawaii) [reply src/YT/YT_GetParticleAttribute.cpp:1]: @hyschive : as shown in the [Slack channel](https://calab-ntu.slack.com/archives/CMJ7B31AT/p1712092380499369?thread_ts=1711873804.671039&cid=CMJ7B31AT), we use `CDM_LSS` test problem slightly modified such that it allows `UM_IC` input, and a periodic `UM_IC` fluid field (with `5` components) with randomly generated `k` numbers is used as input, along with the `PAR_IC` (both `OPT__UM_IC_FLOAT8` and `PAR_IC_FLOAT8` set as `0`). Four combination (`FLOAT8 = 0, 1` ; `FLOAT8_PAR = 0, 1`) and tested, and their outcome seem to be consistent (shown by the movie in the slack thread).

**Comment by koarakawaii** [comment]
@hyschive: the PR has been revised according to the above comment~

**Review by hyschive — APPROVED** [review]
@koarakawaii Looks good. Thanks for the contribution! I'll merge it right away.

**Pull request merged**